go/cbrotli.go

```diff
@@ -0,0 +1,264 @@
+// output can be decoded to match all input before Flush(), but the stream is
+// not yet complete until after Close().
+// Flush has a negative impact on compression.
+func (z *Writer) Flush() error {
+	_, err := z.writeChunk(nil, encoder.Finish)
+	return err
+}
+
+// Encode returns content encoded with Brotli.
+func Encode(content []byte, options WriterOptions) ([]byte, error) {
+	var buf bytes.Buffer
+	writer := NewWriter(&buf, options)
+	defer writer.Close()
+	_, err := writer.Write(content)
+	if err != nil {
+		return nil, err
+	}
+	if err := writer.Close(); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+//------------------------------------------------------------------------------
+// Decoder
+//------------------------------------------------------------------------------
+
+// ErrDecodeFailure is returned the decode operation fails.
+var ErrDecodeFailure = errors.New("Brotli decode failure")
+
+// ErrReadAfterClose indicates Read has been called after Close.
+var ErrReadAfterClose = errors.New("Read() after Close()")
+
+// Reader implements io.ReadCloser, an io.Reader decorator that decodes
+// Brotli-encoded data.
+type Reader struct {
+	src     io.Reader
+	decoder decoder.Decoder
+	buf     []byte // intermediate read buffer pointed to by next
+	eof     bool   // true if all compressed stream is decoded
+	next    []byte // buffered data to be passed to decoder
+	output  []byte // data produced by decoder, but not yet consumed
+	srcErr  error  // last source reader error
+	err     error  // reader state; nil if it is OK to read further
+}
+
+// NewReader initializes new Reader instance.
+// Close MUST be called to free resources.
+func NewReader(src io.Reader) *Reader {
+	return &Reader{
+		src:     src,
+		decoder: decoder.New(),
+		buf:     make([]byte, 32*1024),
+		eof:     false,
+	}
+}
+
+// Close implements io.Closer. Close() MUST be invoked to free native resources.
+func (z *Reader) Close() error {
+	if z.err != ErrReadAfterClose {
+		z.decoder.Close()
+		z.err = ErrReadAfterClose
+	}
+	return nil
+}
+
+func isEOF(src io.Reader) bool {
+	var buf []byte
+	buf = make([]byte, 1)
+	n, err := src.Read(buf)
+	return n == 0 && err == io.EOF
+}
+
+// Read implements io.Reader.
+func (z *Reader) Read(p []byte) (int, error) {
+	// Any error state is unrecoverable.
+	if z.err != nil {
+		return 0, z.err
+	}
+	// See io.Reader documentation.
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	var totalOutBytes int
+	totalOutBytes = 0
+
+	// There is no practical limit for amount of bytes being consumed by decoder
+	// before producing any output. Continue feeding decoder until some data is
+	// produced
+	for {
+		// Push already produced output first.
+		var outBytes int
+		outBytes = len(z.output)
+		if outBytes != 0 {
+			outBytes = copy(p, z.output)
+			p = p[outBytes:]
+			z.output = z.output[outBytes:]
+			totalOutBytes += outBytes
+			// Output buffer is full.
+			if len(p) == 0 {
+				break
+			}
+			continue
```

Comment:
IIUC, hitting this line implies that `len(z.output) == 0` since all of it would have been copied to `p` otherwise? Thus, should we just drop this `continue` and move on in the loop?